pkg/zap/zap.go

```diff
@@ -133,60 +134,61 @@ var levelStrings = map[string]zapcore.Level{
 func zapHook() mapstructure.DecodeHookFuncType {
 	return func(in reflect.Type, out reflect.Type, val interface{}) (interface{}, error) {
 		if in.Kind() == reflect.String && out == levelEnablerType {
-			levelVal := val.(string)
-
+			sVal := val.(string)
 			// return nil if ZAP_LOG_LEVEL is not set, crzap lib sets the default value
-			if levelVal == "" {
+			if sVal == "" {
 				return nil, nil
 			}
 
 			// ZAP_LOG_LEVEL supports setting of integer value > 0 in addition to `info`, `error` or `debug`
-			level, validLevel := levelStrings[strings.ToLower(levelVal)]
+			level, validLevel := levelStrings[strings.ToLower(sVal)]
 			if !validLevel {
-				logLevel, err := strconv.Atoi(levelVal)
+				logLevel, err := strconv.Atoi(sVal)
 				if err != nil {
-					return nil, fmt.Errorf("invalid log level \"%s\"", levelVal)
+					return nil, fmt.Errorf("invalid log level \"%s\"", val)
 				}
 
 				if logLevel > 0 {
 					intLevel := -1 * logLevel
 					return zap.NewAtomicLevelAt(zapcore.Level(int8(intLevel))), nil
 				} else {
-					return nil, fmt.Errorf("invalid log level \"%s\"", levelVal)
+					return nil, fmt.Errorf("invalid log level \"%s\"", val)
 				}
 			}
 
 			return zap.NewAtomicLevelAt(level), nil
 		} else if in.Kind() == reflect.String && out == newEncoderFuncType {
-			// TODO: implement EncodeHook interface for type NewEncoderFunc upstream
-			encoder := crzap.NewEncoderFunc(newJSONEncoder)
-			encoderIn := val.(string)
-
-			if encoderIn == "" {
-				return encoder, nil
-			}
+			// TODO: implement encoding.TextUnmarshaler interface for type NewEncoderFunc upstream
+			var encoder crzap.NewEncoderFunc
 
-			switch encoderIn {
+			switch val.(string) {
+			case "": // Encoder not configured; use default encoder
+				// TODO: Isn't the default encoder dependant on the development/production setting?
+				encoder = newJSONEncoder
```

Comment:
good catch! https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/zap/zap.go#L185 looking at the addDefaults logic, NewEncoder is set correctly based on ZAP_DEVELOPMENT flag, so I think we should set encoder to nil here

Reply:
I think we should create an issue to follow up on this. If setting it to nil, the test fails:

````
Summarizing 1 Failure:
  [FAIL] Zap log level flag options setup with zap-encoding flag provided [It] Should default to json encoder when not set [encoder]
  /home/erikbo/projects/github/controller-runtime-viper/pkg/zap/zap_test.go:241
````

Reply:
Opened issue #51